lib/host/wasi/wasifunc.cpp

```diff
@@ -1525,6 +1524,164 @@ Expect<uint32_t> WasiSchedYield::body(Runtime::Instance::MemoryInstance *) {
   return __WASI_ERRNO_SUCCESS;
 }
 
+Expect<uint32_t> WasiSockOpen::body(Runtime::Instance::MemoryInstance *MemInst,
+                                    int32_t AddressFamily, int32_t SockType,
+                                    uint32_t RoFdPtr) {
+  /// Check memory instance from module.
+  if (MemInst == nullptr) {
+    return __WASI_ERRNO_FAULT;
+  }
+
+  __wasi_fd_t *const RoFd =
+      MemInst->getPointer<__wasi_fd_t *>(RoFdPtr, sizeof(__wasi_fd_t));
+  if (RoFd == nullptr) {
+    return __WASI_ERRNO_FAULT;
+  }
+  const __wasi_address_family_t family =
+      static_cast<__wasi_address_family_t>(AddressFamily);
+  const __wasi_sock_type_t sock_type =
+      static_cast<__wasi_sock_type_t>(SockType);
+  if (auto Res = Env.sockOpen(family, sock_type); unlikely(!Res)) {
+    return Res.error();
+  } else {
+    *RoFd = *Res;
+  }
+
+  return __WASI_ERRNO_SUCCESS;
+}
+
+Expect<uint32_t> WasiSockClose::body(Runtime::Instance::MemoryInstance *MemInst,
+                                     int32_t Fd) {
+  /// Check memory instance from module.
+  if (MemInst == nullptr || Fd < 0) {
+    return __WASI_ERRNO_FAULT;
+  }
+
+  int error = close(Fd);
+  if (unlikely(error < 0)) {
+    return __WASI_ERRNO_FAULT;
+  }
+
+  return __WASI_ERRNO_SUCCESS;
+}
+
+Expect<uint32_t> WasiSockBind::body(Runtime::Instance::MemoryInstance *MemInst,
+                                    int32_t Fd, uint32_t AddressPtr,
+                                    uint32_t Port) {
+  /// Check memory instance from module.
+  if (MemInst == nullptr) {
+    return __WASI_ERRNO_FAULT;
+  }
+  if (Fd < 0) {
+    return __WASI_ERRNO_BADF;
+  }
+
+  __wasi_address_t *InnerAddress = MemInst->getPointer<__wasi_address_t *>(
+      AddressPtr, sizeof(__wasi_address_t));
+
+  if (InnerAddress->buf_len != 4 && InnerAddress->buf_len != 16) {
+    return __WASI_ERRNO_INVAL;
+  }
+
+  unsigned char *AddressBuf = MemInst->getPointer<unsigned char *>(
+      InnerAddress->buf, sizeof(unsigned char) * InnerAddress->buf_len);
+  if (AddressBuf == nullptr) {
+    return __WASI_ERRNO_FAULT;
+  }
+  const __wasi_fd_t WasiFd = Fd;
+
+  if (auto Res = Env.sockBind(WasiFd, AddressBuf, InnerAddress->buf_len,
+                              (uint16_t)Port);
+      unlikely(!Res)) {
+    return Res.error();
+  }
+  return __WASI_ERRNO_SUCCESS;
+}
+
+Expect<uint32_t>
+WasiSockListen::body(Runtime::Instance::MemoryInstance *MemInst, int32_t Fd,
+                     uint32_t Backlog) {
+  /// Check memory instance from module.
+  if (MemInst == nullptr) {
+    return __WASI_ERRNO_FAULT;
+  }
+  if (Fd < 0) {
```

Comment:
No need to check `Fd`

Reply:
Sorry I don't understand why the Fd checking could be done in anothor function

Reply:
Because it's transferred directly to the `Env.sockListen` function. You don't need to check here.

Reply:
got it